### PR TITLE
Add log streaming to the created job

### PR DESCRIPTION
**Pull request by amritghimire**
When running a job from studio, this adds a log streaming which checks
for the updates from the Studio.

Related PR: https://github.com/iterative/studio/pull/11068

image

### CODE REVIEW

**Review by dreadatour — APPROVED** [review]
Looks good to me! Thank you for this update 🥰

File: src/datachain/remote/studio.py [hunk 1]
```diff
@@ -231,6 +235,38 @@ def _unpacker_hook(code, data):
...
+        Follow job logs via websocket connection.
+
+        Args:
+            job_id: ID of the job to follow logs for
+
+        Yields:
+            Dict containing either job status updates or log messages
+        """
+        parsed_url = urlparse(self.url)
+        ws_url = urlunparse(parsed_url._replace(scheme="ws"))
```
Comment (0x2b3bfa0) [inline src/datachain/remote/studio.py:249]: Let's not use WebSockets without TLS, please. At least, let's match the scheme from `self.url` converting `http` to `ws` and `https` to `wss` to simplify local development.

```suggestion
        ws_url = urlunparse(parsed_url._replace(scheme="wss"))
```

**Review by 0x2b3bfa0 — CHANGES_REQUESTED** [review]

File: src/datachain/remote/studio.py [hunk 2]
```diff
@@ -231,6 +235,40 @@ def _unpacker_hook(code, data):
...
+            job_id: ID of the job to follow logs for
+
+        Yields:
+            Dict containing either job status updates or log messages
+        """
+        parsed_url = urlparse(self.url)
+        ws_url = urlunparse(
+            parsed_url._replace(scheme="wss" if parsed_url.scheme == "https" else "ws")
+        )
+        ws_url = f"{ws_url}/logs/follow/?job_id={job_id}&team_name={self.team}"
```
Comment (0x2b3bfa0) [inline src/datachain/remote/studio.py:252]: Why do we need to pass the team name?

**Review by 0x2b3bfa0 — APPROVED** [review]
Still not sure WebSockets are a good idea, especially for log streaming. I'm not even talking about SSE here, but even about plain old polling.

**Review by 0x2b3bfa0 — COMMENTED** [review]
It works, but I wonder if it should be closer to real-time; the following example seems to be buffering the whole output and takes ages to print the logs: 🧐 

#### `datachain studio run test.py`
```python
import time

for number in range(60 * 60):
  time.sleep(0.1)
  print(number)
```

**Comment by amritghimire** [comment]
> It works, but I wonder if it should be closer to real-time; the following example seems to be buffering the whole output and takes ages to print the logs: 🧐 
> 
> 
> 
> #### `datachain studio run test.py`
> 
> ```python
> 
> import time
> 
> 
> 
> for number in range(60 * 60):
> 
>   time.sleep(0.1)
> 
>   print(number)
> 
> ```

Found an possible cause in Studio side. Working on it alongside the test cases for this.

File: src/datachain/remote/studio.py (hunk 2 above)
Comment (amritghimire) [reply src/datachain/remote/studio.py:252]: It seems to be an existing approach for Studio endpoints that takes the team name to check for permission before any other tasks. I reused the same decorator and approach to make sure the permission is correct

Comment (0x2b3bfa0) [reply src/datachain/remote/studio.py:252]: This doesn't feel right to me. 🤔 In theory, the job identifier and the access token should be the only values needed to check for permission; anything else is a potential footgun in the best case, or an actual footgun in the worst case.

Comment (0x2b3bfa0) [reply src/datachain/remote/studio.py:252]: Resolving as off-topic, but pinging @dreadatour just to make noise.

**Pull request merged**